concent_api/core/models.py

```diff
@@ -73,3 +76,167 @@ class MessageAuth(Model):
+        ],
+    }
+
+    class Meta:
+        unique_together = (
+            ('requestor', 'task_id'),
+            ('requestor', 'subtask_id'),
+        )
+
+    task_id                     = models.CharField(max_length = MESSAGE_TASK_ID_MAX_LENGTH)
+    subtask_id                  = models.CharField(max_length = MESSAGE_TASK_ID_MAX_LENGTH, unique = True, db_index = True)
+
+    # Relation to concent client in requestor context.
+    requestor                   = models.ForeignKey(Client, related_name = 'subtasks_as_requestor')
+
+    # Relation to concent client in provider context.
+    provider                    = models.ForeignKey(Client, related_name = 'subtasks_as_provider')
+
+    state                       = models.CharField(max_length = 32, choices = SubtaskState.choices())
+    _current_state              = None
+
+    # If in an active state, it's the time at which Concent automatically transitions to a passive state if it does
+    # not get the information it expects from the client (a timeout). Must be NULL in a passive state.
+    next_deadline               = models.DateTimeField(blank = True, null = True)
+
+    # Related messages
+    task_to_compute             = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_task_to_compute')
+    report_computed_task        = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_report_computed_task')
+    ack_report_computed_task    = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_ack_report_computed_task')
+    reject_report_computed_task = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_reject_report_computed_task')
+    subtask_results_accepted    = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_subtask_results_accepted')
+    subtask_results_rejected    = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_subtask_results_rejected')
+
+    def __init__(self, *args, **kwargs):
+        super(Subtask, self).__init__(*args, **kwargs)
+        if self.pk is not None:
+            self._current_state = self.state
+
+    def clean(self):
+        super(Subtask, self).clean()
+
+        # next_deadline can be None only for passive states
+        if self.pk is not None and self.next_deadline is None and self.state not in self.SUBTASK_PASSIVE_STATES:
+            raise ValidationError({
+                'next_deadline': 'next_deadline is None for active state.'
+            })
+
+        # State transition can happen only by defined rule
+        # but not when we create new object
+        # and not when state is not being changed
+        if (
+            self.pk is not None and
+            self._current_state != self.state and
+            self._current_state not in self.POSSIBLE_STATE_TRANSITIONS[self.state]
```

Comment:
Assuming that someone might try to skip this validation in some ways, and removing it from model for a function actually makes it more possible to someone just don't use the validating function. On model it is at least in a single place, and no one will forget to use it ever.

> Like, it tries to remember the initial value of state but it can't know if it's really the state that is currently in the database. It only knows that the model was created with that initial value. There can be situations where the initial value does not come from the database.

It doesn't if someone will not try to set `pk` manually in constructor. But I have found out there is a recently new feature called `Model.from_db()` which can be overridden and used to remember only state from object that was retrieved from database. It can be combined with check of `self._state.adding` in `save()` to verify if it is already stored yet object.

> So if you for whatever reason just create an instance of Subtask with state set to X and then assign it Y, not because it changed in the table but because you want an instance with a different value, the validation may fail.
> Also, I think it won't even work in all cases. I see that it saves state in constructor. But what if you don't set state via constructor? E.g. you set an initial value by directly assigning it to the field?

It doesn't matter because validation is done when you call `full_clean`. If you change state several times, and call `full_clean` it will still compare with the original backuped state from database.

But off course someone can manipulate `_current_state_name` manually, but as I have said, it is the same scenario as someone is not using validation function at all.

Summing it up, I don't see any benefits from making it a separate function and using it in views. But I see a benefit from having it on model - it will not be forgot to call.

Reply:
My biggest problem with it is that it completely breaks my expectations regarding what validations do. I expect the `clear()` method to be stateless and limited to checking the data in the model. This code tries to track what the programmer does with the instance and it's something outside of its responsibility. The view is responsible for making state transitions not the model.

The problem here (that one might forget about validating the transition) is really why [I suggested creating a separate class responsible for doing transitions](https://github.com/golemfactory/concent/issues/140#issue-299734609) and keeping the transition logic in a single place, out of the view. This is just something that should be checked by code doing the transition, not by the model itself.

Well, if you really, really want to keep it in the model (which I think is not a good separation of responsibilities) you can it in a less hacky way. Instead of tracking user's actions fetch the old version of the model from the database inside `clean()` and compare with that. That's really the value we want to be comparing it with - the value that's currently in the database - not the value that just happened to be passed by someone to the constructor.

Reply:
Please check the updated code - it is not done in a constructor anymore but in a way django recommends. And nothing depends on `pk` which can be set via constructor too. 

If we want this validation to be done in a different place, lets create a separate issue - I don't think it is something I can do quite quickly. But personally I think the responsibilities are ok - it is Subtask which has state, so Subtask is validating the states.